### PR TITLE
Update Tensor doc

### PR DESCRIPTION
**Pull request by albanD**
Add to the Tensor doc info about `.device`, `.is_cuda`, `.requires_grad`, `.is_leaf` and `.grad`.
Update the `register_backward_hook` doc with a warning stating that it does not work in all cases.
Add support in the `_add_docstr` function to add docstring to attributes.

There is an explicit cast here but I am not sure how to handle it properly. The thing is that the doc field for getsetdescr is written as being a const char * (as all other doc fields in descriptors objects) in cpython online documentation. But in the code, it is the only one that is not const.
I assumed here that it is a bug in the code because it does not follow the doc and the convention of the others descriptors and so I cast out the const.
EDIT: the online doc I was looking at is for 3.7 and in that version both the code and the doc are const. For older versions, both are non const.
Please let me know if this should not be done. And if it should be done if there is a cleaner way to do it !

### CODE REVIEW

**Comment by albanD** [comment]
As expected clang tidy is not happy with the cast.

**Comment by ezyang** [comment]
Do a const_cast and that should shut clang tidy up. (I haven't verified if the cast is valid or not, but if you're sure you're right , that's what you should do)

**Comment by albanD** [comment]
I do think the cast is valid here to the best of my knowledge:
Basically, up to python 3.6, the field is not const as you see [here](https://github.com/python/cpython/blob/3.6/Include/descrobject.h#L15) and it becomes const for python 3.7.
In our case I think it's fine to always give it our const tensor as numpy's implementation gives them [here](https://github.com/python/cpython/blob/3.6/Include/descrobject.h#L15) the result of `PyString_AS_STRING` as seen [here](https://github.com/numpy/numpy/blob/1d38e41a2acf67c59a3515af2abecbdbf280dc04/numpy/core/src/multiarray/compiled_base.c#L1397) which must not be modified.

**Comment by albanD** [comment]
The CI fail seems due to a division by 0 in Batchnorm1D test. So I guess unrelated to this PR?

**Review by ezyang — APPROVED** [review]
okey dokey
